### PR TITLE
[Feature] Implement CachingHiveMetastore

### PR DESCRIPTION
**Pull request by stephen-shelby**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #
https://github.com/StarRocks/starrocks/issues/11349

## Problem Summary(Required) ：

- We use guava loading cache as cache structure.
- CachingHiveMetastore could cache any hive metadata used in planner.
- Both Catalog-level and Query-level cache use CachingMetastore as instance. The `expireAfterWrite` and `refreshAfterWrite` of the two is different.
- Query-level cache instance need to reuse catalog-level cache instance to access user's hive metastore.
- I'll rename NewHivePartitionName to HivePartitionName and remove original HivePartitionName After refactoring others related codes

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 115 / 150 (76.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/hive/NewHivePartitionName.java|22|43|51.16%|[52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 63, 64, 84, 93, 109, 110, 111, 112, 113, 114, 115]|
|:large_blue_circle:|com/starrocks/external/hive/CachingHiveMetastore.java|93|107|86.92%|[108, 127, 244, 245, 246, 247, 249, 271, 272, 273, 280, 281, 282, 283]|

File: fe/fe-core/src/main/java/com/starrocks/external/hive/CachingHiveMetastore.java [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+        List<NewHivePartitionName> hivePartitionNames = partitionNames.stream()
+                .map(partitionName -> NewHivePartitionName.of(dbName, tblName, partitionName))
+                .peek(hivePartitionName -> checkState(hivePartitionName.getPartitionNames().isPresent(),
+                        "partition name is missing"))
+                .collect(Collectors.toList());
+
+        Map<NewHivePartitionName, Partition> all = getAll(partitionCache, hivePartitionNames);
+        ImmutableMap.Builder<String, Partition> partitionsByName = ImmutableMap.builder();
+        for (Map.Entry<NewHivePartitionName, Partition> entry : all.entrySet()) {
+            partitionsByName.put(entry.getKey().getPartitionNames().get(), entry.getValue());
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/external/hive/CachingHiveMetastore.java:192]: I will ensure that optional is present at the upper calling code.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 123 / 150 (82.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/hive/NewHivePartitionName.java|30|43|69.77%|[55, 57, 58, 59, 83, 92, 108, 109, 110, 111, 112, 113, 114]|
|:large_blue_circle:|com/starrocks/external/hive/CachingHiveMetastore.java|93|107|86.92%|[108, 127, 244, 245, 246, 247, 249, 271, 272, 273, 280, 281, 282, 283]|

**Review by gengjun-git — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/external/hive/CachingHiveMetastore.java [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+public class CachingHiveMetastore implements IHiveMetastore {
+    public static final long NEVER_CACHE = 0;
+    public static final long NEVER_EVICT = -1;
+    public static final long NEVER_REFRESH = -1;
+    protected final IHiveMetastore metastore;
+
+    protected LoadingCache<String, List<String>> databaseNamesCache;
+    protected LoadingCache<String, List<String>> tableNamesCache;
+    protected LoadingCache<HiveTableName, List<String>> partitionKeysCache;
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/external/hive/CachingHiveMetastore.java:42]: add comment

File: fe/fe-core/src/main/java/com/starrocks/external/hive/CachingHiveMetastore.java [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+    protected LoadingCache<String, Database> databaseCache;
+    protected LoadingCache<HiveTableName, Table> tableCache;
+
+    protected LoadingCache<NewHivePartitionName, Partition> partitionCache;
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/external/hive/CachingHiveMetastore.java:47]: add comment

File: fe/fe-core/src/main/java/com/starrocks/external/hive/CachingHiveMetastore.java [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+                NEVER_REFRESH,
+                perQueryCacheMaxSize,
+                true);
+    }
+
+    protected CachingHiveMetastore(IHiveMetastore metastore, Executor executor, long expireAfterWriteSec,
+                                   long refreshIntervalSec, long maxSize, boolean enableListNamesCache) {
+        this.metastore = metastore;
+
+        if (enableListNamesCache) {
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/external/hive/CachingHiveMetastore.java:65]: add comment, why we need `enableListNamesCache`

File: fe/fe-core/src/main/java/com/starrocks/external/hive/CachingHiveMetastore.java (hunk 3 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/external/hive/CachingHiveMetastore.java:49]: Others could view the definition of class. I think it isn't necessary to write too many comments to introduce these on here

**Comment by DorianZheng** [comment]
run starrocks_admit_test

**Comment by stephen-shelby** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/external/hive/NewHivePartitionName.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    public List<String> getPartitionValues() {
+        return partitionValues;
+    }
+
+    public Optional<String> getPartitionNames() {
+        return partitionNames;
+    }
+
+    public boolean approximateMatchTable(String db, String tblName) {
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/external/hive/NewHivePartitionName.java:82]: what's the usage of this API

File: fe/fe-core/src/main/java/com/starrocks/external/hive/NewHivePartitionName.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    public static NewHivePartitionName of(String dbName, String tblName, List<String> partitionValues) {
+        return new NewHivePartitionName(dbName, tblName, partitionValues);
+    }
+
+    public static NewHivePartitionName of(String dbName, String tblName, String partitionNames) {
+        return new NewHivePartitionName(dbName, tblName, Utils.toPartitionValues(partitionNames), Optional.of(partitionNames));
+    }
+
+    public static List<String> fromPartitionKey(PartitionKey key) {
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/external/hive/NewHivePartitionName.java:49]: the retrun type of `NewHivePartitionName::fromPartitionKey` is not a `NewHivePartitionName`, seems a little confused

**Comment by stephen-shelby** [comment]
run starrocks_fe_unittest

**Comment by stephen-shelby** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 123 / 150 (82.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/hive/NewHivePartitionName.java|30|43|69.77%|[55, 57, 58, 59, 83, 92, 108, 109, 110, 111, 112, 113, 114]|
|:large_blue_circle:|com/starrocks/external/hive/CachingHiveMetastore.java|93|107|86.92%|[112, 130, 247, 248, 249, 250, 252, 274, 275, 276, 283, 284, 285, 286]|

File: fe/fe-core/src/main/java/com/starrocks/external/hive/NewHivePartitionName.java (hunk 5 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/external/hive/NewHivePartitionName.java:82]: I'll comment it in next patch

File: fe/fe-core/src/main/java/com/starrocks/external/hive/NewHivePartitionName.java (hunk 6 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/external/hive/NewHivePartitionName.java:49]: I'll remove it to other class in next patch

**Review by DorianZheng — APPROVED** [review]

**Pull request merged**